lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -275,6 +275,54 @@ struct CircuitLoweringState {
     binds.push_back(op);
   }
 
+  /// For a given named bundle, return the corresponding AliasType. Create and
+  /// record the AliasType, if it doesn't exist for the given named bundle. Can
+  /// only be called for bundles with a valid name.
+  hw::TypeAliasType getTypeAliasForNamedBundle(Type rawType,
+                                               BundleType firrtlType,
+                                               Location typeLoc) {
+
+    auto bundleName = firrtlType.getBundleName();
+    assert(bundleName &&
+           "getTypeAliasForNamedBundle can only be called for a named bundle");
+    // If the lowered type already exists, return that.
+    auto iterT1 = firrtlTypeToAliasTypeMap.find(firrtlType);
+    if (iterT1 != firrtlTypeToAliasTypeMap.end())
+      return iterT1->second;
+    // Mutex, to update the top level Module, and add the TypeDeclOp inside the
+    // TypeDeclOp.
+    std::lock_guard<std::mutex> lock(typeScopeMutex);
+    // The typescope is created only at the top level module.
+    Block &topLevelBlock = circuitOp->getParentRegion()->getBlocks().back();
+    if (!typeScope) {
+      // Initialize typeScope.
+      auto b = ImplicitLocOpBuilder::atBlockBegin(circuitOp.getLoc(),
+                                                  &topLevelBlock);
+      typeScope = b.create<hw::TypeScopeOp>(
+          b.getStringAttr(circuitOp.getName() + "__TYPESCOPE_"));
+      typeScope.getBodyRegion().push_back(new Block());
+    }
+    // Get a unique typedecl name.
+    // The bundleName can conflict with other symbols, but must be unique within
+    // the TypeScopeOp.
+    bundleName =
+        StringAttr::get(bundleName.getContext(),
+                        typeDeclNamespace.newName(bundleName.getValue()));
+
+    auto typeScopeBuilder =
+        ImplicitLocOpBuilder::atBlockEnd(typeLoc, typeScope.getBodyBlock());
+    auto typeDecl = typeScopeBuilder.create<hw::TypedeclOp>(typeLoc, bundleName,
```

Comment:
typeLoc is going to vary wildly between minor changes in optimization or circuit.  It's also not useful.